src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs

```diff
@@ -0,0 +1,361 @@
+
+                case MoveAction.Down:
+                    return GetElements(unconfiguredProject, project, projectTree).LastOrDefault();
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Move a child element based on the reference element and move action.
+        /// </summary>
+        private static void MoveChildByMoveAction(ProjectItemElement element, ProjectItemElement referenceElement, MoveAction moveAction)
+        {
+            var parent = referenceElement.Parent;
+            if (parent != null)
+            {
+                parent.RemoveChild(element);
+
+                switch (moveAction)
+                {
+                    case MoveAction.Up:
+                        parent.InsertBeforeChild(element, referenceElement);
+                        break;
+
+                    case MoveAction.Down:
+                        parent.InsertAfterChild(element, referenceElement);
+                        break;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Moves child elements based on the reference element and move action.
+        /// </summary>
+        private static void MoveChildrenByMoveAction(IEnumerable<ProjectItemElement> elements, ProjectItemElement referenceElement, MoveAction moveAction)
+        {
+            // If we are moving elements down, we need to reverse order.
+            if (moveAction == MoveAction.Down)
+            {
+                elements = elements.Reverse();
+            }
+
+            foreach (var element in elements)
+            {
+                MoveChildByMoveAction(element, referenceElement, moveAction);
+            }
+        }
+
+        /// <summary>
+        /// Move project elements based on the given project tree and move action. 
+        /// Will modify the project if successful, but not save; only dirty.
+        /// </summary>
+        private static bool TryMove(UnconfiguredProject unconfiguredProject, Project project, IProjectTree projectTree, MoveAction moveAction)
+        {
+            // Determine what sibling we want to look at based on if we are moving up or down.
+            var sibling = GetSiblingByMoveAction(projectTree, moveAction);
+
+            if (sibling != null)
+            {
+                // The reference element is the element we want to insert above or below.
+                var referenceElement = GetReferenceElementByMoveAction(unconfiguredProject, project, sibling, moveAction);
+
+                if (referenceElement != null && referenceElement.Parent != null)
+                {
+                    // If the project tree we are wanting to move is a folder, we might do many insertions and removals.
+                    if (projectTree.IsFolder)
+                    {
+                        var elements = GetElements(unconfiguredProject, project, projectTree);
+                        if (elements.Count() == 0)
```

Comment:
Prefer `!elements.Any()` rather than forcing a linear walk of the whole collection if it isn't something that is type tested for a Count/Length.